### PR TITLE
Added program and hook prototypes

### PR DESCRIPTION
**Pull request by trishms**
Defined program types, attach types, and callouts. Set up program info and data for registration, unregistration, for future hook work. Created skeleton for callout functions. Added generated program headers which will contain encoded program info.

### CODE REVIEW

File: netebpfext/net_ebpf_ext.c [hunk 1]
```diff
@@ -195,6 +253,46 @@ static net_ebpf_ext_wfp_callout_state_t _net_ebpf_ext_wfp_callout_state[] = {
...
+        L"Flow Established V4 Callout",
+        L"Flow Established callout driver for eBPF",
+        FWP_ACTION_CALLOUT_TERMINATING,
+    },
+    {
+        &EBPF_HOOK_FLOW_ESTABLISHED_CALLOUT_V6,
+        &FWPM_LAYER_ALE_FLOW_ESTABLISHED_V6,
+        _net_ebpf_ext_flow_established_classify,
+        _net_ebpf_ext_no_op_notify,
+        _net_ebpf_ext_flow_delete,
```
Comment (keith-horton) [inline netebpfext/net_ebpf_ext.c:274]: we don't want to know when the flows are deleted?

File: include/ebpf_nethooks.h [hunk 2]
```diff
@@ -78,7 +78,7 @@ bind_hook_t(bind_md_t* context);
 
 
 // Five-tuple Struct
-typedef struct five_tuple
+typedef struct _five_tuple
 {
     bool v4;
```
Comment (lianghaoGao) [inline include/ebpf_nethooks.h:83]: does this compile with only stdint.h?

File: libs/api_common/windows_platform_common.cpp [hunk 3]
```diff
@@ -44,6 +44,18 @@ const ebpf_context_descriptor_t g_bind_context_descriptor = {
 const EbpfProgramType windows_bind_program_type =
     PTYPE("bind", &g_bind_context_descriptor, (uint64_t)&EBPF_PROGRAM_TYPE_BIND, {"bind"});
 
+const ebpf_context_descriptor_t g_flow_context_descriptor = {
+    sizeof(flow_md_t), 0, EBPF_OFFSET_OF(flow_md_t, app_id) + sizeof(uint64_t), -1};
```
Comment (lianghaoGao) [inline libs/api_common/windows_platform_common.cpp:48]: why the offset of app_id? Same question for packet_length.

File: netebpfext/net_ebpf_ext.c [hunk 4]
```diff
@@ -43,48 +43,67 @@
...
     EBPF_OFFSET_OF(xdp_md_t, data_end),
     EBPF_OFFSET_OF(xdp_md_t, data_meta)};
 static ebpf_program_info_t _ebpf_xdp_program_info = {{"xdp", &_ebpf_xdp_context_descriptor, {0}}, 0, NULL};
-
 static ebpf_program_data_t _ebpf_xdp_program_data = {&_ebpf_xdp_program_info, NULL};
-
 static ebpf_extension_data_t _ebpf_xdp_program_info_provider_data = {
     NET_EBPF_EXTENSION_NPI_PROVIDER_VERSION, sizeof(_ebpf_xdp_program_data), &_ebpf_xdp_program_data};
 
+// Bind Hook
```
Comment (lianghaoGao) [inline netebpfext/net_ebpf_ext.c]: unnecessary change

File: netebpfext/net_ebpf_ext.c [hunk 5]
```diff
@@ -651,6 +749,20 @@ net_ebpf_ext_register_providers()
         return STATUS_UNSUCCESSFUL;
     }
 
+    return_value = ebpf_ext_attach_register_provider(
```
Comment (lianghaoGao) [inline netebpfext/net_ebpf_ext.c:819]: Are you missing an input argument?

File: netebpfext/net_ebpf_ext.c [hunk 6]
```diff
@@ -144,9 +179,32 @@ _net_ebpf_ext_resource_release_classify(
     uint64_t flow_context,
     _Inout_ FWPS_CLASSIFY_OUT* classify_output);
 
+static void
+_net_ebpf_ext_flow_established_classify(
```
Comment (lianghaoGao) [inline netebpfext/net_ebpf_ext.c:186]: Missing function definition

File: netebpfext/net_ebpf_ext.c [hunk 7]
```diff
@@ -144,9 +179,32 @@ _net_ebpf_ext_resource_release_classify(
...
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output);
+
 static void
 _net_ebpf_ext_no_op_flow_delete(uint16_t layer_id, uint32_t fwpm_callout_id, uint64_t flow_context);
 
+static void
+_net_ebpf_ext_flow_delete(uint16_t layer_id, uint32_t fwpm_callout_id, uint64_t flow_context);
```
Comment (lianghaoGao) [inline netebpfext/net_ebpf_ext.c:209]: missing function definition

File: include/ebpf_nethooks.h (hunk 2 above)
Comment (trishms) [reply include/ebpf_nethooks.h:83]: There is <stdbool.h> as well and it does compile!

File: netebpfext/net_ebpf_ext.c (hunk 1 above)
Comment (trishms) [reply netebpfext/net_ebpf_ext.c:274]: We do, that will be implemented in the _net_ebpf_ext_flow_delete function. Will be implementing it afterwards. I only have the skeleton for the callouts right now.

File: libs/api_common/windows_platform_common.cpp (hunk 3 above)
Comment (trishms) [reply libs/api_common/windows_platform_common.cpp:48]: Since it is the last variable of the struct, I assumed that its offset + its size would be the offset of the pointer to end of data?

File: netebpfext/net_ebpf_ext.c (hunk 4 above)
Comment (trishms) [reply netebpfext/net_ebpf_ext.c]: Did you mean the comments and spacing changes?

File: netebpfext/net_ebpf_ext.c (hunk 6 above)
Comment (trishms) [reply netebpfext/net_ebpf_ext.c:186]: Thanks, I added the function definition in my local repo, will push these changes after implementing the other feedback. I will add the logic into the functions once I start implementing the flow established and mac hooks.

File: netebpfext/net_ebpf_ext.c (hunk 5 above)
Comment (trishms) [reply netebpfext/net_ebpf_ext.c:819]: The input arguments are there (next line)!

File: libs/api_common/windows_platform_common.cpp (hunk 3 above)
Comment (lianghaoGao) [reply libs/api_common/windows_platform_common.cpp:48]: You don't have any pointer types in your struct, right? Can you make sure it passes verifier?

File: netebpfext/net_ebpf_ext.c (hunk 4 above)
Comment (lianghaoGao) [reply netebpfext/net_ebpf_ext.c]: yes

File: netebpfext/net_ebpf_ext.c (hunk 6 above)
Comment (lianghaoGao) [reply netebpfext/net_ebpf_ext.c:186]: Are you planning to add those implementation with a seperate PR?

File: netebpfext/net_ebpf_ext.c (hunk 5 above)
Comment (lianghaoGao) [reply netebpfext/net_ebpf_ext.c:819]: I'm saying ebpf_ext_attach_register_provider() requires 4 input parameters but you only have 3 here. Does this even compile?

File: netebpfext/net_ebpf_ext.c (hunk 6 above)
Comment (trishms) [reply netebpfext/net_ebpf_ext.c:186]: I will add it to this PR, I just wanted to get your feedback (like the comment and spacing one) and so I can implement that too before I push again.

File: netebpfext/net_ebpf_ext.c (hunk 5 above)
Comment (trishms) [reply netebpfext/net_ebpf_ext.c:819]: image
It does compile. I think there is only 3 inputs.

Comment (lianghaoGao) [reply netebpfext/net_ebpf_ext.c:819]: Maybe I guess this forked repo doesn't have the latest changes from ebpf-for-windows proj. Does it matter?

**Review by lianghaoGao — APPROVED** [review]
